service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CreateTableCopyJobsStep.java

```diff
@@ -0,0 +1,194 @@
+import com.google.api.services.bigquery.model.TableReference;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateTableCopyJobsStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(CreateTableCopyJobsStep.class);
+  public static final Duration COPY_JOB_TIMEOUT = Duration.ofHours(12);
+  private final CrlService crlService;
+  private final WorkspaceService workspaceService;
+  private final ControlledBigQueryDatasetResource sourceDataset;
+
+  public CreateTableCopyJobsStep(
+      CrlService crlService,
+      WorkspaceService workspaceService,
+      ControlledBigQueryDatasetResource sourceDataset) {
+    this.crlService = crlService;
+    this.workspaceService = workspaceService;
+    this.sourceDataset = sourceDataset;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    final FlightMap workingMap = flightContext.getWorkingMap();
+    // Gather inputs
+    final DatasetCloneInputs sourceInputs = getSourceInputs();
+    workingMap.put(ControlledResourceKeys.SOURCE_CLONE_INPUTS, sourceInputs);
+
+    final DatasetCloneInputs destinationInputs = getDestinationInputs(flightContext);
+    workingMap.put(ControlledResourceKeys.DESTINATION_CLONE_INPUTS, destinationInputs);
+
+    final AuthenticatedUserRequest userRequest =
+        flightContext
+            .getInputParameters()
+            .get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    final BigQueryCow bigQueryCow = crlService.createBigQueryCow(userRequest);
+    // TODO:  remove usage of this client when it's all in CRL
+    final Bigquery bigQueryClient = crlService.createNakedBigQueryClient(userRequest);
+    try {
+      // Get a list of all tables in the source dataset
+      final TableList sourceTables =
+          bigQueryCow
+              .tables()
+              .list(sourceInputs.getProjectId(), sourceInputs.getDatasetName())
+              .execute();
+      // Start a copy job for each source table
+      final Map<String, String> tableToJobId =
+          Optional.ofNullable(
+                  workingMap.get(
+                      ControlledResourceKeys.TABLE_TO_JOB_ID_MAP,
+                      new TypeReference<Map<String, String>>() {}))
+              .orElseGet(HashMap::new);
+      for (TableList.Tables table : sourceTables.getTables()) {
+        if (tableToJobId.containsKey(table.getId())) {
+          // A job already exists for this table
+          continue;
+        }
+        // If the table contains data in its streaming buffer, that data can't be copied yet
+        // https://cloud.google.com/bigquery/streaming-data-into-bigquery#dataavailability
```

Comment:
TIL, that's an interesting interaction.
It seems like if the buffer is not empty, we could either copy whatever we can, or throw an error. Do we have a strong case for either one?

Reply:
I think the case where there is pending streaming input happening AND a clone happening is unlikely.
The main use case for clone is copying a featured workspace. Those won’t have streaming data pouring in.
My suggestion is that we document this behavior and move on.

Reply:
I agree that this is a BigQuery limitation, which can also be seen in the Console, and we don't need to try to do anything about it but warn folks.

I still kinda want to tack this warning message onto the output somehow, but it seems like the restructuring necessary would be ugly.